lib/backends/sdc/affinity.js

```diff
@@ -0,0 +1,722 @@
+ */
+
+/* BEGIN JSSTYLED */
+/*
+ * Copyright 2016 Joyent, Inc.
+ *
+ * Container affinity support (i.e. the rules/hints for deciding to what
+ * server a new container is provisioned).
+ *
+ * The goal is to provide the affinity features that Docker Swarm provides
+ * with its "affinity" container filters, described here:
+ *      https://docs.docker.com/swarm/scheduler/filter/#how-to-write-filter-expressions
+ * The other Swarm filters are ignored. See DOCKER-630 for discussion.
+ *
+ * # Affinity types
+ *
+ * There are three affinity axes in the Swarm docs:
+ *
+ * - *container affinity*: Specify to land on the same or different server
+ *   as an existing container(s).
+ *      docker run -e affinity:container==db0 ...
+ *      docker run --label 'com.docker.swarm.affinities=["container==db0"]' ...
+ *
+ * - *label affinity*: Specify to land on the same or different server as
+ *   existing containers with a given label key/value.
+ *      docker run --label role=webhead ...     # the starter container
+ *      docker run -e affinity:role==webhead ...
+ *      docker run --label 'com.docker.swarm.affinities=["role==webhead"]' ...
+ *
+ * - *image affinity*: Specify to land on a node with the given image.
+ *      docker run -e affinity:image==redis ...
+ *      docker run --label 'com.docker.swarm.affinities=["image==redis"]' ...
+ *   Note: We will skip this one. For Triton an image is present on all nodes
+ *   in the DC. Until a possible future when Triton acts as a Swarm master
+ *   for multiple DCs, the semantics of this affinity don't apply.
+ *
+ * # Affinities -> Locality Hints
+ *
+ * Triton's feature for a VM creation providing affinity is "locality hints".
+ * Therefore we'll be translating given Docker affinities (via both
+ * the '-e' envvar syntax and the newer '--label' syntax) to Triton's
+ * "locality hints". See here for the locality hints big-theory comment
+ * and implementation:
+ *      https://github.com/joyent/sdc-designation/blob/master/lib/algorithms/soft-filter-locality-hints.js
+ *
+ * # Limitations
+ *
+ * - sdc-designation's locality hints cannot handle mixed strict and non-strict
+ *   rules. E.g.:
+ *      docker run -e affinity:container==db0 -e 'affinity:container!=db1' ...
+ *   To support that we'd need to extend the "locality" data structure format.
+ *   Currently we just drop the non-strict rules when hitting this. An
+ *   alternative would be to error out.
+ */
+/* END JSSTYLED */
+
+var assert = require('assert-plus');
+var format = require('util').format;
+var strsplit = require('strsplit');
+var vasync = require('vasync');
+var XRegExp = require('xregexp');
+
+var common = require('../../common');
+var errors = require('../../errors');
+
+
+// ---- globals
+
+var FILTER_KEY_RE = /^[a-z_][a-z0-9\-_.]+$/i;
+// Note that this differs from the Swarm one in expr.go to fix some issues
+// and accomodate slight parsing differences.
+// JSSTYLED
+var FILTER_VALUE_RE = /^[-a-z0-9:_\s.*/()?+[\]\\^$|]+$/i;
```

Comment:
Might be good to list which types of values this is supposed to allow. Regexes are not the most readable.